src/pmu.rs

```diff
@@ -0,0 +1,301 @@
+            self.pmusleep.write(|w| w.sleep().set_bit());
+        }
+    }
+
+    fn store_backup<UD>(&self, user_data: &UD) -> Result<(), BackupError>
+    where
+        UD: Sized,
+    {
+        unsafe {
+            let backup = BACKUP::ptr();
+            let ud_size = core::mem::size_of::<UD>();
+
+            if ud_size > (*backup).backup.len() * BACKUP_REGISTER_BYTES {
+                return Err(BackupError::DataTooLarge);
+            }
+
+            if ud_size % BACKUP_REGISTER_BYTES != 0 {
+                return Err(BackupError::DataSizeInvalid);
+            }
+
+            let reg_count = ud_size / BACKUP_REGISTER_BYTES;
+
+            let ptr = user_data as *const _;
+            let ptr_u32 = ptr as *const u32;
+            let sliced = core::slice::from_raw_parts(ptr_u32, reg_count);
+
+            for i in 0..sliced.len() {
+                (*backup).backup[i].write(|w| w.bits(sliced[i]));
+            }
+
+            Ok(())
+        }
+    }
+
+    fn restore_backup<UD>(&self, user_data: UD) -> Result<UD, BackupError>
+    where
+        UD: Sized,
+    {
+        unsafe {
+            let backup = BACKUP::ptr();
+            let ud_size = core::mem::size_of::<UD>();
+
+            if ud_size > (*backup).backup.len() * BACKUP_REGISTER_BYTES {
+                return Err(BackupError::DataTooLarge);
+            }
+
+            if ud_size % BACKUP_REGISTER_BYTES != 0 {
+                return Err(BackupError::DataSizeInvalid);
+            }
+
+            let reg_count = ud_size / BACKUP_REGISTER_BYTES;
+
+            let ptr = &user_data as *const _;
+            let ptr_u32 = ptr as *mut u32;
+            let sliced = core::slice::from_raw_parts_mut(ptr_u32, reg_count);
+
+            for i in 0..sliced.len() {
+                sliced[i] = (*backup).backup[i].read().bits();
+            }
+
+            Ok(user_data)
+        }
+    }
+
+    fn clear_backup(&self) {
+        unsafe {
+            let backup = BACKUP::ptr();
+
+            for i in 0..(*backup).backup.len() {
+                (*backup).backup[i].write(|w| w.bits(0u32));
+            }
+        }
+    }
+
+    fn wakeup_cause(&self) -> Result<WakeupCause, CauseError> {
+        let pmu_cause = self.pmucause.read();
+        let wakeup_cause = pmu_cause.wakeupcause();
+        if wakeup_cause.is_rtc() {
+            return Ok(WakeupCause::RTC)
+        } else if wakeup_cause.is_digital() {
+            return Ok(WakeupCause::Digital)
+        } else if wakeup_cause.is_reset() {
+            let reset_cause = pmu_cause.resetcause();
+
+            if reset_cause.is_power_on() {
+                return Ok(WakeupCause::Reset(ResetCause::PowerOn))
+            } else if reset_cause.is_external() {
+                return Ok(WakeupCause::Reset(ResetCause::External))
+            } else if reset_cause.is_watchdog() {
+                return Ok(WakeupCause::Reset(ResetCause::WatchDog))
+            }
+        }
+
+        Err(CauseError::InvalidCause)
```

Comment:
I think that `WakeupCause::Unknown` or `WakeupCause::Other(x)` will match this condition better.

Reply:
I wanted this to fail explicitly in case something new is added on the board in the future. What should a user do if they get unknown? Error out anyway, this way I suspect they might be more inclined to report a bug.